Add client list endpoint

PgrNascimento opened the pull request:
## Objective

Map [Client List endpoint](https://documenter.getpostman.com/view/4377973/RW8FDkDM?version=latest#697aaf7a-3a60-8d5b-bdbc-6e44b4633efd) from Glassfactory API and resolve the https://github.com/Cendrao/glass_factory_api/issues/8 issue

## TODO

- [x] Add documentation

## Bonus

- Update the get_client and get_client! functions docs

PgrNascimento committed:
Add matches when the API Client returns an error

It's possible that the API Client returns errors other than those
reported. So I've added one match for tuple `{:ok, a body with error}`
and other match for `{:error, any error}`

PgrNascimento committed:
Add list_client/1 function

This function returns a list of clients for an account. In this first
version, the client just have: id, name, archived_at, owner_id and
office_id

PgrNascimento committed:
Add list_clients!/1 function

This function works like list_client\1, but here we don't use a tuple
when we received a success response and raises when the return
from the API Client is an error.

PgrNascimento committed:
Add docs to list_clients and list_clients! functions

PgrNascimento committed:
Update the get_client and get_client! function docs

paulo-silva commented on lib/clients.ex:
```diff
@@ -6,33 +6,138 @@ defmodule GlassFactoryApi.Clients do
+    with {:ok, %{status: 200, body: body}} <- ApiClient.get("clients", config) do
+      {:ok, Enum.map(body, &Client.to_struct(&1))}
+    else
+      {:error, error} when is_atom(error) -> {:error, Atom.to_string(error)}
```
I think that we don't need it anymore, considering that we handle it on ApiClient 🤔

PgrNascimento replied:
@paulo-silva Could you give me more context about it?

Cendrao replied:
After [this PR](https://github.com/Cendrao/glass_factory_api/pull/27/files), we are handling that kind of error in `ApiClient` as you can see [here](https://github.com/Cendrao/glass_factory_api/blob/c0498aa11ad70b0dccba80bf53a907c7ff8fa025/lib/api_client.ex#L21-L26).

So you can ✂️  this guard and just handle `{:error, error}` where `error` will always be a string now.

PgrNascimento replied:
Thank you guys <3

PgrNascimento replied:
Done here: https://github.com/Cendrao/glass_factory_api/pull/25/commits/54819f4a4c6d7d8ff18d8f7a9df4a6d06a322de5

britto commented on lib/clients.ex:
```diff
@@ -6,33 +6,138 @@ defmodule GlassFactoryApi.Clients do
-        owner_id: 567,
-        office_id: 789
-      }
+  iex> GlassFactoryApi.Clients.get_client("1234")
```
The code samples are inlined so that the documentation tools can syntax-highlight them correctly. Example from Elixir itself: https://github.com/elixir-lang/elixir/blob/055526057ac9992cc1df87441357c4e56d45235d/lib/elixir/lib/string.ex#L1186-L1200

PgrNascimento replied:
My editor alined the examples in a wrong way :(

I've done the suggestion here: https://github.com/Cendrao/glass_factory_api/pull/25/commits/b9602fb6001f00bc82a1f0f3907ca523cbf4d7ad

Thanks <3

britto commented on lib/clients.ex:
```diff
@@ -6,33 +6,138 @@ defmodule GlassFactoryApi.Clients do
     else
       {:ok, %{status: 404}} -> {:error, "Can't find a client with id #{client_id}"}
+      {:ok, %{body: body}} -> {:error, body}
+      {:error, error} -> {:error, error}
```
Will Tesla guarantee `error` is a string here?

PgrNascimento replied:
In this moment, I don't know if Tesla guarantee for us the return is a String, @Cendrao could you help me this one? 😊

Cendrao replied:
Tesla won't guarantee that, but I think `ApiClient` does, because of [this implementation](https://github.com/Cendrao/glass_factory_api/blob/c0498aa11ad70b0dccba80bf53a907c7ff8fa025/lib/api_client.ex#L21-L26).

PgrNascimento committed:
Apply @britto sugestions

Use the correct indentation for functions examples

PgrNascimento committed:
Merge branch 'master' of github.com:Cendrao/glass_factory_api into aon-implement-list-clients

* 'master' of github.com:Cendrao/glass_factory_api:
  Normalize Members resource (#28)
  Apply @Cendrao suggestion.
  Remove not more necessary transformation.
  Normalize ApiClient result.
  Create GlassFactoryApi.Clients.Reports.list_time_reports!/2.
  Create GlassFactoryApi.Clients.Reports.list_time_reports/2.
  Add TimeReport struct.

PgrNascimento committed:
Remove unnecessary guard_clause

After the PR#27, when the Tesla API returns an atom and not a string,
the ApiClient module treats it and returns a String. So, this guard
clause now is unnecessary.

Cendrao reviewed (approved):
:shipit:

PgrNascimento committed:
Merge branch 'master' of github.com:Cendrao/glass_factory_api into aon-implement-list-clients

* 'master' of github.com:Cendrao/glass_factory_api:
  Ignore tool-versions file
